agent/operator_endpoint.go

```diff
@@ -104,6 +105,17 @@ func (s *HTTPServer) OperatorKeyringEndpoint(resp http.ResponseWriter, req *http
 		}
 	}
 
+	// Parse local-only. local-only will only be used in GET requests.
+	if localOnly := req.URL.Query().Get("local-only"); localOnly != "" {
+		var err error
+		args.LocalOnly, err = strconv.ParseBool(localOnly)
+		if err != nil {
+			resp.WriteHeader(http.StatusBadRequest)
```

Comment:
Could you make this return a `BadRequestError{Reason: fmt.Sprintf("Error parsing local-only: %v", err)} instead of writing to the request here.

The story about this goes way back but eventually we are going to want to clean up our error responses in the API. Passing back the error type instead of writing the response here will allow us to do that. We are a long ways away from that goal but when implementing new code we can use the error types to minimize future work.

Reply:
yes no problem! i like that much better too! 

q: if we are returning an `error`, it's probably a good idea to maintain the actual error instance instead of just the fmt.Sprintf + err.Error()? With the way error handling is moving in the Go community (folks wanting to wrap errors, add context, etc, and the error itself holding stuff like stack trace) I feel like this would be a good practice to start getting into?
It looks like maybe that would be a bigger project, but perhaps adding an `error` field to all the http `Error` structs would be a good start?
LMK what you think! (also, maybe there is a better forum for talking about this?)

Reply:
I'll make your change in the meantime though @mkeeler

Reply:
Hm, actually maybe this should be done in a different PR? there are a bunch of other instances of this error handling style in this function, and in this file. I'd be happy to submit a follow up PR cleaning up all of those. LMK thoughts - sorry for all the replies.

Reply:
Sounds good to me @adams-sarah!

Reply:
awesome, thx @i0rek

Reply:
I love the idea of having strongly typed errors instead of just passing strings around. That is a much larger undertaking so probably outside of scope for here.

Reply:
Are you volunteering to overhaul the existing API to do this? ;)

Thats quite an undertaking. I normally take the approach of not introducing more "bad" stuff into the code (there is probably a better adjective to use there). So in this case I would just make it return the BadRequestError and shelve the rest until someone (maybe you) wants to go through the whole HTTP API and make things consistent.

Reply:
re: errors - awesome! i'll lodge this in the back of my brain for later
re: overhaul - i worry about this inconsistency in terms of unforeseen user-facing behavior inconsistencies between a group of api endpoints in the way errors are handled and return. [with no knowledge of this particular project] i would rather see consistency on this at least in a file or group of endpoints, rather than one one-off change in error handling. particularly since this change is to pass off dealing with this error to another function, and not just to deal with it differently right here. and i'm happy to make this change for this file! :). wdyt?

Reply:
Doing it for this whole file would be great if your willing.

For a little more context:

If you return a random error from one of the HTTP endpoint funcs then it gets translated into a 503. In many, many places its user request issues and not an internal server issue. For that reason the BadRequestError was created so that we don't have to set the response code and write the output all over the place. We still do all of that its just done in a common place to reduce code redundancy.

I personally would worry more about consistency of the APIs when each one implements the error reporting on their own. I see each time we don't do that but send an error back up the chain as a win for ensuring that all our APIs function similarly with regards to errors. 

I personally take the more incremental approach of improving the situation as I haven't had a whole lot of time to focus on big tech debt things like this. This does have the downside of potentially breaking things if someone were to modify how the errors are changed within http.go but for that to happen someone would have to have reviewed the code and missed the change.

I don't think I would hold up the PR over it as you seem to have your own contrary opinions to mine and its a really minor thing anyways but I do still believe even the localized change would be good.

Reply:
ok! it's not a huge deal for me either.
how about i'll just make the changes for errors here in the PR as you suggested, and then also make a follow-on PR (as I have lots of time!) to change the rest of the instances in these files.

💩 hard consistency just sits better in my brain is all. thanks for the conversation :)